### PR TITLE
Adds support for Tunnel Dweller, Vehicle Collisions and Heavy Scientist

### PR DESCRIPTION
**Pull request by johnmcclumpha**
Adds support for Tunnel Dwellers and vehicle collisions.

Requires `CombatEntityTypes.json` to have tunnel dwellers added as:
```  "tunneldweller": 17```
(Note: I wasn't sure of the conventions being used for the values here, so 17 may or may not be "correct")

Facilitates the messaging as per examples below in `config/DeathNotes.json`:

```
      {
        "KillerType": "TunnelDweller",
        "VictimType": "Player",
        "DamageType": "*",
        "Messages": [
          "{victim} was taken out by a {killer}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "TunnelDweller",
        "DamageType": "*",
        "Messages": [
          "{killer} took out a {victim} with their {weapon}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "Animal",
        "DamageType": "Collision",
        "Messages": [
          "{killer} ran over a {victim}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "Player",
        "DamageType": "Collision",
        "Messages": [
          "{killer} ran over {victim}."
        ]
      },
```

### CODE REVIEW

**Comment by jeremyjohn** [comment]
Thank You!

**Comment by bnagy1** [comment]
Im not sure what I did wrong, I added the plugin but can get tunnel dwellers to show up when killed. then I tried to add  the following to json file but no luck? am I doing somthing wrong?

 {
        "KillerType": "TunnelDweller",
        "VictimType": "Player",
        "DamageType": "*",
        "Messages": [
          "{victim} was taken out by a {killer}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "TunnelDweller",
        "DamageType": "*",
        "Messages": [
          "{killer} took out a {victim} with their {weapon}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "Animal",
        "DamageType": "Collision",
        "Messages": [
          "{killer} ran over a {victim}."
        ]
      },
      {
        "KillerType": "Player",
        "VictimType": "Player",
        "DamageType": "Collision",
        "Messages": [
          "{killer} ran over {victim}."
        ]
      },

**Comment by Snugglebit** [comment]
I doesn't work for me either.
Whenever I add "tunneldweller": 17 to the CombatEntityTypes.json and reload the plugin, the entry disappears again. What do I need to do, so that it stays?

**Comment by johnmcclumpha** [comment]
> I doesn't work for me either.
> Whenever I add "tunneldweller": 17 to the CombatEntityTypes.json and reload the plugin, the entry disappears again. What do I need to do, so that it stays?

@Snugglebit This is intended as a patch for the plugin - for it to function without the remote configs being updated you'll need to override the loading of them. There's a separate branch over here that will do this: https://github.com/johnmcclumpha/DeathNotes/blob/temp/prevents-remote-loading/DeathNotes.cs

**Comment by MrKraken** [comment]
For the new underwater update, do the same as above but for `underwaterdweller`, again i just followed numerically and made `CombatEntityTypes.json` to have:
`"underwaterdweller": 18`

**Comment by cary-williams** [comment]
Works great for me except for one thing. For some reason with this patch, zombies from the ZombieHorde plugin now show as Heavy Scientists in the death notes. Console still shows correct though `Zombie[7486685] was killed by Nemesys`

I think it may just be something weird with the way the author spawns them in, but im not sure. I notice also that before the patch, they show as zombie but it uses the default message of `Nemesys did not want to be a part of the Zombie's experiments` same as a scientist, despite have a specific zombie message set in the config and DefaultMessages.json

**Comment by eldubya** [comment]
@MrKraken Your instructions work perfectly, thank you. Any ideas on how to add the new Simple Shark as well? I have been messing around with the files but can't get it working.

File: DeathNotes.cs
Comment (LaserHydra) [inline DeathNotes.cs:19]: Please don't increment the version yet. I will do that once before merging develop into master.
(I just created the develop branch now)

File: DeathNotes.cs
Comment (LaserHydra) [inline DeathNotes.cs:389]: These are redundant, as the last line in this method should already handle this.
Notably though, the same is the case for the existing Helicopter case above. You can also remove that please.

File: DeathNotes.cs
Comment (LaserHydra) [inline DeathNotes.cs:396]: This won't be necessary here, as I've added the Heavy Scientist to the Enemy Prefabs dictionary.
For that to become effective, I will have to move that statement from below above this switch-statement, but I will do that after your changes are through.

File: DeathNotes.cs
Comment (LaserHydra) [inline DeathNotes.cs:439]: I would move the TunnelDweller entry to below the Scarecrow entry perhaps, as it makes a bit more sense there and to keep Other and None at the bottom.

It also looks like the TunnelDweller should have a switch-case in GetEntityName, returning "Tunnel Dweller".

File: DeathNotes.cs
Comment (LaserHydra) [inline DeathNotes.cs:520]: Please adjust the brace placement to follow the conventions used here, i.e. move it to the next line.

**Review by LaserHydra — CHANGES_REQUESTED** [review]
Thank you for your contributions and apologies for my (very) late response.
I've reviewed your changes and noted a few changes to make.

I also applied the change to the CombatEntityTypes file. However, to keep backward compatibility, I've versioned the files.

So please, also adjust the RemoteConfiguration Host to the following, so the changes become effective.
```http://files.laserhydra.com/config/DeathNotes/v6.3.6/```

**Comment by Timmooo** [comment]
Has there been any updates from this? Is it just waiting on @johnmcclumpha to complete the changes?

**Comment by LaserHydra** [comment]
@Timmooo It's been a little off the radar for me while waiting for a response.
I'll try to pick this up again in the next couple of days and adjust it myself.

**Comment by eldubya** [comment]
Any progress on this @LaserHydra ? Sorry to keep on about it, but it's a great plugin that I'm sure is used by almost every modded server out there, myself included.

**Comment by LaserHydra** [comment]
@eldubya 
Sorry, I've had a bunch of things on my plate and haven't gotten around to it yet.
I'll keep this tab open to remind me every day though, when I get a chance.

**Comment by MrGraversen** [comment]
We appreciate the effort @LaserHydra

**Comment by LaserHydra** [comment]
I have now updated Death Notes with these changes and a few more small things.
https://umod.org/plugins/death-notes#updates

Thank you for your contributions again, nonetheless @johnmcclumpha

**Pull request closed without merging**
